File: examples/typescript/react/notes-showcase/src/viewmodels/capabilityActionsVM.ts
/**
 * CapabilityActionsVM — projects a focused VM's capability surface into a
 * flat list of `ActionVM`s for the capability-action bar (spec §14.4).
 *
 * VMx-API adaptation: TypeScript's interfaces are structural, but the
 * capability registry (CAP-020) requires explicit `declareCapabilities`
 * markers — we use `hasCapability` instead of `instanceof` checks for
 * capability discovery. NoteVM (the only example-defined target of
 * `IDeletable<NoteVM>` / `ISavable<NoteVM>`) is detected by `instanceof`
 * so we can wire the right argument.
 */
import { BehaviorSubject, map } from "rxjs";
import {
  ComponentVMBase,
  DerivedProperty,
  hasCapability,
  RelayCommand,
  ViewModelType,
  type ICommand,
  type IDispatcher,
  type IMessageHub,
} from "@thekaveh/vmx";

import { type ActionVM, makeActionVM } from "./actionVM.js";
import { NoteVM } from "./noteVM.js";

type FocusedGetter = () => object | null;

interface SelectionShape {
  canSelect(): boolean;
  select(): void;
}
interface DeselectionShape {
  canDeselect(): boolean;
  deselect(): void;
}
interface ToggleSelectionShape {
  canToggleSelection(): boolean;
  toggleSelection(): void;
}
interface ExpandShape {
  canExpand(): boolean;
  expand(): void;
}
interface CollapseShape {
  canCollapse(): boolean;
  collapse(): void;
}
interface ToggleExpansionShape {
  canToggleExpansion(): boolean;
  toggleExpansion(): void;
}
interface ClosableShape {
  canClose(): boolean;
  close(): void;
}
interface ApprovableShape {
  canApprove(): boolean;
  approve(): void;
}
interface CancelableShape {
  canCancel(): boolean;
  cancel(): void;
}
interface NewCreatableShape {
  canCreateNew(): boolean;
  createNew(): void;
}
interface ReconstructShape {
  canReconstruct(): boolean;
  reconstruct(): void;
}

export class CapabilityActionsVM extends ComponentVMBase {
  readonly #focusedGetter: FocusedGetter;
  readonly #focusSubject: BehaviorSubject<object | null>;
  readonly #actionsDerived: DerivedProperty<readonly ActionVM[]>;
  // Edge-case backfill (readonly notebook gating): a dedicated *Add Note*
  // command whose `canExecute` predicate consults the host-supplied
  // `canAddNote` callback. The host (e.g. `WorkspaceVM`) wires this to
  // `!notesView.currentNotebookIsReadonly`. Defaults stay backward-compatible:
  // a no-op action and an always-true predicate.
  readonly #addNoteCommand: RelayCommand;

  constructor(opts: {
    name: string;
    hint: string;
    hub: IMessageHub;
    dispatcher: IDispatcher;
    focusedGetter: FocusedGetter;
    addNoteAction?: () => void;
    canAddNote?: () => boolean;
  }) {
    super({
      name: opts.name,
      hint: opts.hint,
      hub: opts.hub,
      dispatcher: opts.dispatcher,
    });
    this.#focusedGetter = opts.focusedGetter;
    this.#focusSubject = new BehaviorSubject<object | null>(
      opts.focusedGetter(),
    );
    this.#actionsDerived = new DerivedProperty<readonly ActionVM[]>(
      this.#focusSubject.pipe(map((f) => CapabilityActionsVM.project(f))),
      null,
      null,
    );
    const addNoteAction = opts.addNoteAction ?? ((): void => undefined);
    const canAddNote = opts.canAddNote ?? ((): boolean => true);
    this.#addNoteCommand = RelayCommand.builder()
      .predicate(canAddNote)
      .task(addNoteAction)
      .build();
  }

  get type(): ViewModelType {
    return ViewModelType.Component;
  }

  get hub(): IMessageHub {
    return this._hub;
  }

  get focusedVM(): object | null {
    return this.#focusedGetter();
  }

  /** Live projection: list of (label, command) for the focused VM. */
  get actions(): DerivedProperty<readonly ActionVM[]> {
    return this.#actionsDerived;
  }

  /**
   * Host-driven *Add Note* command.
   *
   * Edge-case backfill (readonly notebook gating): `canExecute` delegates
   * to the `canAddNote` callback supplied at construction. `WorkspaceVM`
   * wires the predicate to `!notesView.currentNotebookIsReadonly` so the
   * bar gates adding notes to readonly notebooks.
   */
  get addNoteCommand(): ICommand {
    return this.#addNoteCommand;
  }

  /**
   * Re-projects the action list from the current focused VM. The host calls
   * this after focus changes (typically via a hub subscription).
   */
  recomputeActions(): void {
    this.#focusSubject.next(this.#focusedGetter());
  }

  static project(focused: object | null): readonly ActionVM[] {
    if (focused === null) return [];
    const out: ActionVM[] = [];

    if (hasCapability(focused, "ISelectable")) {
      const f = focused as SelectionShape;
      out.push(
        makeActionVM(
          "Select",
          RelayCommand.builder()
            .predicate(() => f.canSelect())
            .task(() => f.select())
            .build(),
        ),
      );
    }
    if (hasCapability(focused, "IDeselectable")) {
      const f = focused as DeselectionShape;
      out.push(
        makeActionVM(
          "Deselect",
          RelayCommand.builder()
            .predicate(() => f.canDeselect())
            .task(() => f.deselect())
            .build(),
        ),
      );
    }
    if (hasCapability(focused, "ISelectionTogglable")) {
      const f = focused as ToggleSelectionShape;
      out.push(
        makeActionVM(
          "Toggle Selection",
          RelayCommand.builder()
            .predicate(() => f.canToggleSelection())
            .task(() => f.toggleSelection())
            .build(),
        ),
      );
    }
    if (hasCapability(focused, "IExpandable")) {
      const f = focused as ExpandShape;
      out.push(
        makeActionVM(
          "Expand",
          RelayCommand.builder()
            .predicate(() => f.canExpand())
            .task(() => f.expand())
            .build(),
        ),
      );
    }
    if (hasCapability(focused, "ICollapsible")) {
      const f = focused as CollapseShape;
      out.push(
        makeActionVM(
          "Collapse",
          RelayCommand.builder()
            .predicate(() => f.canCollapse())
            .task(() => f.collapse())
            .build(),
        ),
      );
    }
    if (hasCapability(focused, "IExpansionTogglable")) {
      const f = focused as ToggleExpansionShape;
      out.push(
        makeActionVM(
          "Toggle Expansion",
          RelayCommand.builder()
            .predicate(() => f.canToggleExpansion())
            .task(() => f.toggleExpansion())
            .build(),
        ),
      );
    }
    if (hasCapability(focused, "IClosable")) {
      const f = focused as ClosableShape;
      out.push(
        makeActionVM(
          "Close",
          RelayCommand.builder()
            .predicate(() => f.canClose())
            .task(() => f.close())
            .build(),
        ),
      );
    }
    if (hasCapability(focused, "IApprovable")) {
      const f = focused as ApprovableShape;
      out.push(
        makeActionVM(
          "Approve",
          RelayCommand.builder()
            .predicate(() => f.canApprove())
            .task(() => f.approve())
            .build(),
        ),
      );
    }
    if (hasCapability(focused, "ICancelable")) {
      const f = focused as CancelableShape;
      out.push(
        makeActionVM(
          "Cancel",
          RelayCommand.builder()
            .predicate(() => f.canCancel())
            .task(() => f.cancel())
            .build(),
        ),
      );
    }
    if (hasCapability(focused, "INewCreatable")) {
      const f = focused as NewCreatableShape;
      out.push(
        makeActionVM(
          "New",
          RelayCommand.builder()
            .predicate(() => f.canCreateNew())
            .task(() => f.createNew())
            .build(),
        ),
      );
    }
    // Save / Delete target NoteVM (the only example-defined VM implementing
    // ISavable<T> / IDeletable<T>). Scenario §6.2: each note saves /
    // deletes itself. Reuse `note.saveCommand` / `note.deleteCommand`
    // directly so the action-bar Delete fires the same
    // ConfirmationDecoratorCommand (and "Note deleted" notification) that
    // the in-list delete button uses — keeping the action-bar and the
    // in-list delete button behaviorally identical (parity with C#
    // CapabilityActionsVM.cs:121-131).
    if (focused instanceof NoteVM) {
      const note = focused;
      if (hasCapability(focused, "ISavable")) {
        out.push(makeActionVM("Save", note.saveCommand));
      }
      if (hasCapability(focused, "IDeletable")) {
        out.push(makeActionVM("Delete", note.deleteCommand));
      }
    }
    if (hasCapability(focused, "IReconstructable")) {
      const f = focused as ReconstructShape;
      out.push(
        makeActionVM(
          "Reconstruct",
          RelayCommand.builder()
            .predicate(() => f.canReconstruct())
            .task(() => f.reconstruct())
            .build(),
        ),
      );
    }
    return out;
  }

  protected override _onDispose(): void {
    this.#actionsDerived.dispose();
    this.#addNoteCommand.dispose();
    this.#focusSubject.complete();
    super._onDispose();
  }

  static builder(): CapabilityActionsVMBuilder {
    return new CapabilityActionsVMBuilder();
  }
}

export class CapabilityActionsVMBuilder {
  #name: string | null = null;
  #hint = "";
  #hub: IMessageHub | null = null;
  #dispatcher: IDispatcher | null = null;
  #focusedGetter: FocusedGetter | null = null;
  #addNoteAction: (() => void) | null = null;
  #canAddNote: (() => boolean) | null = null;

  constructor(from?: CapabilityActionsVMBuilder) {
    if (from) {
      this.#name = from.#name;
      this.#hint = from.#hint;
      this.#hub = from.#hub;
      this.#dispatcher = from.#dispatcher;
      this.#focusedGetter = from.#focusedGetter;
      this.#addNoteAction = from.#addNoteAction;
      this.#canAddNote = from.#canAddNote;
    }
  }

  name(value: string): CapabilityActionsVMBuilder {
    const b = new CapabilityActionsVMBuilder(this);
    b.#name = value;
    return b;
  }

  hint(value: string): CapabilityActionsVMBuilder {
    const b = new CapabilityActionsVMBuilder(this);
    b.#hint = value;
    return b;
  }

  services(
    hub: IMessageHub,
    dispatcher: IDispatcher,
  ): CapabilityActionsVMBuilder {
    const b = new CapabilityActionsVMBuilder(this);
    b.#hub = hub;
    b.#dispatcher = dispatcher;
    return b;
  }

  focusedGetter(getter: FocusedGetter): CapabilityActionsVMBuilder {
    const b = new CapabilityActionsVMBuilder(this);
    b.#focusedGetter = getter;
    return b;
  }

  /** Wire the *Add Note* command body (edge-case backfill). */
  addNoteAction(action: () => void): CapabilityActionsVMBuilder {
    const b = new CapabilityActionsVMBuilder(this);
    b.#addNoteAction = action;
    return b;
  }

  /**
   * Wire the *Add Note* command can-execute predicate.
   *
   * Used by `WorkspaceVM` to consult `notesView.currentNotebookIsReadonly`
   * so the bar disables *Add Note* against readonly notebooks.
   */
  canAddNote(predicate: () => boolean): CapabilityActionsVMBuilder {
    const b = new CapabilityActionsVMBuilder(this);
    b.#canAddNote = predicate;
    return b;
  }

  build(): CapabilityActionsVM {
    if (this.#name === null) throw new Error("name is required");
    if (this.#hub === null || this.#dispatcher === null)
      throw new Error("services are required");
    if (this.#focusedGetter === null)
      throw new Error("focusedGetter is required");
    return new CapabilityActionsVM({
      name: this.#name,
      hint: this.#hint,
      hub: this.#hub,
      dispatcher: this.#dispatcher,
      focusedGetter: this.#focusedGetter,
      ...(this.#addNoteAction !== null
        ? { addNoteAction: this.#addNoteAction }
        : {}),
      ...(this.#canAddNote !== null ? { canAddNote: this.#canAddNote } : {}),
    });
  }
}

File: examples/python/textual/notes_showcase/src/notes_showcase/viewmodels/capability_actions_vm.py
"""CapabilityActionsVM — projects a focused VM's capabilities to ActionVMs.

Inspects the currently-focused VM (supplied by a host-side getter) and
projects every capability interface it implements into a flat list of
:class:`~notes_showcase.viewmodels.action_vm.ActionVM` for the action-bar.
See spec §14.4 (capability dispatch).
"""

from __future__ import annotations

import dataclasses
from collections.abc import Awaitable, Callable

from reactivex.subject import BehaviorSubject

from vmx import (
    AsyncRelayCommand,
    ComponentVM,
    DerivedProperty,
    IApprovable,
    ICancelable,
    IClosable,
    ICollapsible,
    IDeletable,
    IDeselectable,
    IExpandable,
    IExpansionTogglable,
    INewCreatable,
    IReconstructable,
    ISavable,
    ISelectable,
    ISelectionTogglable,
    MessageHub,
    MessageHubProto,
    RelayCommand,
    RxDispatcher,
    from_sources,
)
from vmx.messages.protocols import Message
from vmx.services.dispatcher import Dispatcher

from notes_showcase.viewmodels.action_vm import ActionVM
from notes_showcase.viewmodels.note_vm import NoteVM


class CapabilityActionsVM(ComponentVM):
    """Action-bar VM whose ``actions`` derives from a focused VM's capabilities."""

    def __init__(
        self,
        *,
        name: str,
        hint: str,
        hub: MessageHubProto[Message],
        dispatcher: Dispatcher,
        focused_getter: Callable[[], object | None],
        add_note_action: Callable[[], Awaitable[None] | None] | None = None,
        can_add_note: Callable[[], bool] | None = None,
    ) -> None:
        super().__init__(name=name, hint=hint, hub=hub, dispatcher=dispatcher)
        self._focused_getter = focused_getter
        self._focus_subject: BehaviorSubject[object | None] = BehaviorSubject(
            focused_getter()
        )
        self._actions: DerivedProperty[list[ActionVM]] = from_sources(
            self._focus_subject,
            transform=_project,
        )
        # Edge-case backfill (readonly notebook gating): a dedicated
        # "Add Note" command whose ``can_execute`` predicate consults the
        # host-supplied :meth:`can_add_note` callback. The host (e.g.
        # :class:`WorkspaceVM`) typically wires this to
        # ``not notes_view.current_notebook_is_readonly``. Defaults stay
        # backward-compatible: a no-op action and an always-true predicate.
        self._add_note_action_cb = add_note_action or (lambda: None)
        self._can_add_note_cb: Callable[[], bool] = can_add_note or (lambda: True)
        self._add_note_command: AsyncRelayCommand = (
            AsyncRelayCommand.builder()
            .predicate(self._can_add_note_cb)
            .task(self._invoke_add_note)
            .build()
        )

    # ── Public surface ─────────────────────────────────────────────────────
    @property
    def hub(self) -> MessageHubProto[Message]:
        return self._hub

    @property
    def focused_vm(self) -> object | None:
        return self._focused_getter()

    @property
    def actions(self) -> DerivedProperty[list[ActionVM]]:
        return self._actions

    @property
    def add_note_command(self) -> AsyncRelayCommand:
        """Host-driven "Add Note" command.

        Edge-case backfill (readonly notebook gating): ``can_execute``
        delegates to the ``can_add_note`` callback supplied at construction.
        :class:`WorkspaceVM` wires the predicate to
        ``not notes_view.current_notebook_is_readonly`` so the bar gates
        adding notes to readonly notebooks.
        """
        return self._add_note_command

    async def _invoke_add_note(self) -> None:
        pending = self._add_note_action_cb()
        if pending is not None:
            await pending

    def recompute_actions(self) -> None:
        """Refresh the projection from the focus-getter delegate."""
        self._focus_subject.on_next(self._focused_getter())

    # ── Lifecycle ──────────────────────────────────────────────────────────
    def _on_dispose(self) -> None:
        self._actions.dispose()
        self._add_note_command.dispose()
        self._focus_subject.on_completed()
        self._focus_subject.dispose()
        super()._on_dispose()

    # ── Builder entry-point ────────────────────────────────────────────────
    @staticmethod
    def builder() -> CapabilityActionsVMBuilder:  # type: ignore[override]
        # Narrows ComponentVM.builder() to the showcase CapabilityActionsVMBuilder.
        return CapabilityActionsVMBuilder()


def _project(focused: object | None) -> list[ActionVM]:
    if focused is None:
        return []
    actions: list[ActionVM] = []

    # Selection
    if isinstance(focused, ISelectable):
        actions.append(
            ActionVM(
                "Select",
                RelayCommand.builder()
                .predicate(focused.can_select)
                .task(focused.select)
                .build(),
            )
        )
    if isinstance(focused, IDeselectable):
        actions.append(
            ActionVM(
                "Deselect",
                RelayCommand.builder()
                .predicate(focused.can_deselect)
                .task(focused.deselect)
                .build(),
            )
        )
    if isinstance(focused, ISelectionTogglable):
        actions.append(
            ActionVM(
                "Toggle Selection",
                RelayCommand.builder()
                .predicate(focused.can_toggle_selection)
                .task(focused.toggle_selection)
                .build(),
            )
        )

    # Expansion
    if isinstance(focused, IExpandable):
        actions.append(
            ActionVM(
                "Expand",
                RelayCommand.builder()
                .predicate(focused.can_expand)
                .task(focused.expand)
                .build(),
            )
        )
    if isinstance(focused, ICollapsible):
        actions.append(
            ActionVM(
                "Collapse",
                RelayCommand.builder()
                .predicate(focused.can_collapse)
                .task(focused.collapse)
                .build(),
            )
        )
    if isinstance(focused, IExpansionTogglable):
        actions.append(
            ActionVM(
                "Toggle Expansion",
                RelayCommand.builder()
                .predicate(focused.can_toggle_expansion)
                .task(focused.toggle_expansion)
                .build(),
            )
        )

    # Dialog
    if isinstance(focused, IClosable):
        actions.append(
            ActionVM(
                "Close",
                RelayCommand.builder()
                .predicate(focused.can_close)
                .task(focused.close)
                .build(),
            )
        )
    if isinstance(focused, IApprovable):
        actions.append(
            ActionVM(
                "Approve",
                RelayCommand.builder()
                .predicate(focused.can_approve)
                .task(focused.approve)
                .build(),
            )
        )
    if isinstance(focused, ICancelable):
        actions.append(
            ActionVM(
                "Cancel",
                RelayCommand.builder()
                .predicate(focused.can_cancel)
                .task(focused.cancel)
                .build(),
            )
        )

    # CRUD
    if isinstance(focused, INewCreatable):
        actions.append(
            ActionVM(
                "New",
                RelayCommand.builder()
                .predicate(focused.can_create_new)
                .task(focused.create_new)
                .build(),
            )
        )
    # ISavable / IDeletable are parameterised — only NoteVM implements them
    # in this scenario, and the parameter is always the focused note itself.
    # Reuse ``note.save_command`` / ``note.delete_command`` directly so the
    # action-bar Delete fires the same ConfirmationDecoratorCommand (and
    # "Note deleted" notification) the in-list delete button uses — keeping
    # the action-bar and the in-list delete button behaviorally identical
    # (parity with C# CapabilityActionsVM.cs:121-131).
    if isinstance(focused, NoteVM):
        note: NoteVM = focused
        if isinstance(focused, ISavable):
            actions.append(ActionVM("Save", note.save_command))
        if isinstance(focused, IDeletable):
            actions.append(ActionVM("Delete", note.delete_command))

    # Lifecycle
    if isinstance(focused, IReconstructable):
        actions.append(
            ActionVM(
                "Reconstruct",
                RelayCommand.builder()
                .predicate(focused.can_reconstruct)
                .task(focused.reconstruct)
                .build(),
            )
        )

    return actions


@dataclasses.dataclass(frozen=True, slots=True)
class CapabilityActionsVMBuilder:
    """Immutable fluent builder for :class:`CapabilityActionsVM`."""

    _name: str | None = None
    _hint: str = ""
    _hub: MessageHubProto[Message] | None = None
    _dispatcher: Dispatcher | None = None
    _focused_getter: Callable[[], object | None] | None = None
    _add_note_action: Callable[[], Awaitable[None] | None] | None = None
    _can_add_note: Callable[[], bool] | None = None

    def name(self, value: str) -> CapabilityActionsVMBuilder:
        return dataclasses.replace(self, _name=value)

    def hint(self, value: str) -> CapabilityActionsVMBuilder:
        return dataclasses.replace(self, _hint=value)

    def services(
        self, hub: MessageHubProto[Message], dispatcher: Dispatcher
    ) -> CapabilityActionsVMBuilder:
        return dataclasses.replace(self, _hub=hub, _dispatcher=dispatcher)

    def focused_getter(
        self, getter: Callable[[], object | None]
    ) -> CapabilityActionsVMBuilder:
        return dataclasses.replace(self, _focused_getter=getter)

    def add_note_action(
        self, action: Callable[[], Awaitable[None] | None]
    ) -> CapabilityActionsVMBuilder:
        """Wire the *Add Note* command body (edge-case backfill)."""
        return dataclasses.replace(self, _add_note_action=action)

    def can_add_note(self, predicate: Callable[[], bool]) -> CapabilityActionsVMBuilder:
        """Wire the *Add Note* command can-execute predicate.

        Used by :class:`WorkspaceVM` to consult
        ``notes_view.current_notebook_is_readonly`` so the bar disables
        *Add Note* against readonly notebooks.
        """
        return dataclasses.replace(self, _can_add_note=predicate)

    def build(self) -> CapabilityActionsVM:
        if self._name is None:
            raise ValueError("name is required")
        if self._focused_getter is None:
            raise ValueError("focused_getter is required")
        hub = self._hub if self._hub is not None else MessageHub[Message]()
        dispatcher = (
            self._dispatcher
            if self._dispatcher is not None
            else RxDispatcher.immediate()
        )
        return CapabilityActionsVM(
            name=self._name,
            hint=self._hint,
            hub=hub,
            dispatcher=dispatcher,
            focused_getter=self._focused_getter,
            add_note_action=self._add_note_action,
            can_add_note=self._can_add_note,
        )
